concurrency/TransportBuddyManager.java

```diff
@@ -0,0 +1,270 @@
+     */
+    public synchronized void storePendingStatus(JID jid, PresenceType presence, String status) {
+        if (!isActivated()) {
+            pendingPresences.put(jid, presence);
+            pendingVerboseStatuses.put(jid, status);
+        }
+        else {
+            try {
+                B buddy = getBuddy(jid);
+                buddy.setPresenceAndStatus(presence, status);
+            }
+            catch (NotFoundException e) {
+                // Alrighty then....
+            }
+        }
+    }
+
+    /**
+     * Resets (clears) the list of buddies assigned to the buddy manager.
+     */
+    public void resetBuddies() {
+        buddies.clear();
+    }
+
+    /**
+     * Retrieve the buddy instance for a given JID.
+     *
+     * @param jid JID of the buddy to be retrieved.
+     * @throws NotFoundException if the given jid is not found.
+     * @return TransportBuddy instance requested.
+     */
+    public B getBuddy(JID jid) throws NotFoundException {
+        B buddy = buddies.get(jid);
+        if (buddy == null) {
+            throw new NotFoundException("Could not find buddy requested.");
+        }
+        return buddy;
+    }
+    
+    /**
+     * Retrieve the buddy instance for a given user.
+     *
+     * @param username Username of the buddy to be retrieved.
+     * @throws NotFoundException if the given username is not found.
+     * @return TransportBuddy instance requested.
+     */
+// Disabling this because it can cause confusion in one's attempt to match a buddy.
+//    public B getBuddy(String username) throws NotFoundException {
+//        B buddy = buddies.get(username.toLowerCase());
+//        if (buddy == null) {
+//            throw new NotFoundException("Could not find buddy requested.");
+//        }
+//        return buddy;
+//    }
+
+    /**
+     * Stores a new buddy instance in the buddy list.
+     *
+     * @param buddy TransportBuddy associated with the username.
+     */
+    public void storeBuddy(B buddy) {
+        if (!buddies.containsKey(buddy.jid)) {
+            Log.debug("("+getSession().getTransport().getType().toString().toUpperCase()+") Storing new buddy: "+buddy);
+            buddies.put(buddy.jid, buddy);
+            if (isActivated()) {
+                getSession().lockRoster(buddy.getJID().toString());
+                try {
+                    getSession().getTransport().addOrUpdateRosterItem(getSession().getJID(), buddy.getJID(), buddy.getNickname(), buddy.getGroups());
+                }
+                catch (UserNotFoundException e) {
+                    Log.error("TransportBuddyManager: Unable to find roster when adding contact.");
+                } finally {
+                    getSession().unlockRoster(buddy.getJID().toString());
+                }
+            }
+        }
+        else {
+            Log.debug("("+getSession().getTransport().getType().toString().toUpperCase()+") Replacing buddy: "+buddy);
+            buddies.put(buddy.jid, buddy);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **174 and 191**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).